### PR TITLE
Bug 2217 - Convert DB to use flat schema

### PR DESCRIPTION
**Pull request by jbuck**
This changes projects to be a flat schema, and removes customData.

### CODE REVIEW

File: cornfield/lib/user.js [hunk 1]
```diff
@@ -90,25 +59,27 @@ module.exports = {
...
-    this.findAllProjects( email, function( err, doc ) {
+    Project2Model.find( { email: email, _id: pid }, function( err, doc ) {
       if ( err ) {
         callback( err );
         return;
       }
 
-      callback( err, doc.projects.id( pid ) );
+      // .find() returns an array, but this API expects a single document or null
+      doc = doc.length > 0 ? doc[0] : null;
```
Comment (secretrobotron) [inline cornfield/lib/user.js]: Spacing around the 0.

File: cornfield/migrate.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+ProjectModel = mongoose.model( 'Project', Project ),
+
+User = new Schema({
+  email: String,
+  projects: [Project],
+}),
+UserModel = mongoose.model( 'User', User );
+
+function migrateData() {
+  console.log("Attempting data migration");
```
Comment (secretrobotron) [inline cornfield/migrate.js]: Spacing.

File: cornfield/migrate.js [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  name: String,
+  data: String,
+  template: String,
+  customData: String
+}),
+ProjectModel = mongoose.model( 'Project', Project ),
+
+User = new Schema({
+  email: String,
+  projects: [Project],
```
Comment (secretrobotron) [inline cornfield/migrate.js]: Spacing.

File: cornfield/migrate.js [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+}),
+UserModel = mongoose.model( 'User', User );
+
+function migrateData() {
+  console.log("Attempting data migration");
+
+  UserModel.find( {}, function( err, users ) {
+
+    if ( users.length === 0 ) {
+      console.log("No users to migrate");
```
Comment (secretrobotron) [inline cornfield/migrate.js]: Spacing.

File: cornfield/lib/user.js [hunk 5]
```diff
@@ -90,25 +59,27 @@ module.exports = {
...
+      doc = doc.length > 0 ? doc[ 0 ] : null;
+      callback( err, doc );
     });
   },
   isDBOnline: function isDBOnline() {
     return dbOnline;
   },
   updateProject: function updateProject( email, pid, data, callback ) {
-    if ( !email || !pid || !data || !data.id ) {
+    if ( !email || !pid || !data ) {
```
Comment (mjschranz) [inline cornfield/lib/user.js:77]: I'll just ask because I haven't been able to figure it out. Why do we not need to check for the id anymore?

Comment (jbuck) [inline cornfield/lib/user.js:77]: I use the pid parameter given in the function

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Received

Command **cmd_check** from @jbuck received. Current queue size: 0

Live output at: http://botio.butyr.org:8000/fafa25c85c5bc2c/output.txt

**Comment by ghost** [comment]
#### From: Bot.io (butyr.org)

---

# Success

Full output at http://botio.butyr.org:8000/fafa25c85c5bc2c/output.txt

Total script time: 2.59 mins

**Pull request merged**